sgpt/handlers/handler.py

```diff
@@ -121,12 +130,14 @@ def get_completion(
                 )
                 if tool_calls:
                     for tool_call in tool_calls:
+                        if tool_call.id:
+                            id = tool_call.id
                         if tool_call.function.name:
                             name = tool_call.function.name
                         if tool_call.function.arguments:
                             arguments += tool_call.function.arguments
```

Comment:
It seems a single string concat is not enough if there are multiple tool requirements.
arguments might be """
{"shell_command": "uname -a"}{"shell_command": "lsb_release -a"}
"""
which is not a valid json string.

Reply:
```python
arguments_map = {}
name_map = {}
if tool_calls:
    for tool_call in tool_calls:
        if tool_call.id is not None:
            id = tool_call.id
            name_map[id] = tool_call.function.name
            arguments_map[id] = ""
        else:
            arguments_map[id] += tool_call.function.arguments  
if chunk.choices[0].finish_reason == "tool_calls":
    for id, name, arguments in zip(name_map.keys(), name_map.values(), arguments_map.values()):
        yield from self.handle_function_call(messages, id, name, arguments)
```

This worked for me using deepseek, but not tested in other settings.

Reply:
yes, i forgot about parallel function call, this code also works for me!
![image](https://github.com/user-attachments/assets/7dec42e3-876c-473b-b5ea-ad9ee3b3c067)
maybe anyone have access to gpt-4o can verify this also

Reply:
got it fixed, i think this also a new feature as originally only single function call is support by the code

Reply:
My code assumes there to be multiple chunks in tool_calls, first of which carries id and name, while others streaming tokens of arguments. This might not work for non-streaming mode.  Haven't tried though.

Reply:
yes, but i wonder the code used in shell_gpt only support stream mode, and i guess may be handle multiple function calls in handle_function_call sgpt/handlers/handler.py could be better
```
    def handle_function_call(
        self,
        messages: List[dict[str, Any]],
        id: str,
        name: str,
        arguments: str,
    ) -> Generator[str, None, None]:
        messages.append(
            {
                "role": "assistant",
                "content": "",
                "function_call": {"name": name, "arguments": arguments},
                "tool_calls": [
                    {
                        "id": id,
                        "type": "function",
                        "function": {"name": name, "arguments": arguments},
                    }
                ],
            }
        )
def handle_function_call(
        result = get_function(name)(**dict_args)
        if cfg.get("SHOW_FUNCTIONS_OUTPUT") == "true":
            yield f"```text\n{result}\n```\n"
```
as we can put all function call in "tool_calls" and for iterate call them yield result